source/host/KReplaySystem.cc

```diff
@@ -0,0 +1,209 @@
+}
+
+void KReplaySystem::reportFail(const char *msg) const {
+    std::string report(m_currentGhostFileName);
+    report += "\n" + std::string(msg);
+    Abstract::File::Append("results.txt", report.c_str(), report.size());
+}
+
+bool KReplaySystem::success() const {
+    const auto *raceManager = System::RaceManager::Instance();
+    if (raceManager->stage() != System::RaceManager::Stage::FinishGlobal) {
+        reportFail("Race didn't finish");
+        return false;
+    }
+
+    s32 desyncingTimerIdx = getDesyncingTimerIdx();
+    if (desyncingTimerIdx != -1) {
+        char msgBuffer[128];
+
+        const auto [correct, incorrect] = getDesyncingTimer(desyncingTimerIdx);
+        if (desyncingTimerIdx == 0) {
+            snprintf(msgBuffer, sizeof(msgBuffer),
+                    "Final timer desync! Expected [%d:%d:%d], got [%d:%d:%d]", correct.min,
+                    correct.sec, correct.mil, incorrect.min, incorrect.sec, incorrect.mil);
+        } else {
+            snprintf(msgBuffer, sizeof(msgBuffer),
+                    "Lap %d timer desync! Expected [%d:%d:%d], got [%d:%d:%d]", desyncingTimerIdx,
+                    correct.min, correct.sec, correct.mil, incorrect.min, incorrect.sec,
+                    incorrect.mil);
+        }
+        reportFail(msgBuffer);
+        return false;
+    }
+
+    return true;
+}
+
+/// @brief Finds the desyncing timer index, if one exists.
+/// @return -1 if there's no desync, 0 if the final timer desyncs, and 1+ if a lap timer desyncs.
+s32 KReplaySystem::getDesyncingTimerIdx() const {
+    const auto &player = System::RaceManager::Instance()->player();
+    if (m_currentGhost->raceTimer() != player.raceTimer()) {
+        return 0;
+    }
+
+    for (size_t i = 0; i < 3; ++i) {
+        if (m_currentGhost->lapTimer(i) != player.getLapSplit(i + 1)) {
+            return i + 1;
+        }
+    }
+
+    return -1;
+}
+
+/// @brief Gets the desyncing timer according to the index.
+/// @param i Index to the desyncing timer. Cannot be -1.
+/// @return The pair of timers. The first is the correct one, and the second is the incorrect one.
+std::pair<const System::Timer &, const System::Timer &> KReplaySystem::getDesyncingTimer(
+        s32 i) const {
+    auto cond = i <=> 0;
+    ASSERT(cond != std::strong_ordering::less);
+
+    if (cond == std::strong_ordering::equal) {
+        const auto &correct = m_currentGhost->raceTimer();
+        const auto &incorrect = System::RaceManager::Instance()->player().raceTimer();
+        ASSERT(correct != incorrect);
+        return std::pair<const System::Timer &, const System::Timer &>(correct, incorrect);
+    } else if (cond == std::strong_ordering::greater) {
+        const auto &correct = m_currentGhost->lapTimer(i - 1);
+        const auto &incorrect = System::RaceManager::Instance()->player().lapTimer(i - 1);
+        ASSERT(correct != incorrect);
+        return std::pair<const System::Timer &, const System::Timer &>(correct, incorrect);
```

Comment:
```suggestion
        return std::make_pair(correct, incorrect);
```
Need to make sure it keeps it as a reference though

Reply:
`make_pair` (and subsequently, implicit `std::pair`) will expand to `std::pair<System::Timer, System::Timer>`, which is not what we want.

If this looks ugly, maybe the solution is a private `using` statement.

Reply:
Saw this somewhere, does this work?
```suggestion
        return std::make_pair(std::ref(correct), std::ref(incorrect));
```

Reply:
Nevermind, const-ness would not be kept i think